src/javascript/note.js

```diff
@@ -1,94 +1,215 @@
-export default function Note(vm, x, width, color) {
-  this.x = x;
-  this.width = width;
-  this.color = color;
-  const { ctx, canvas } = vm;
+export default class Note {
+  constructor(vm, game, keyObj, key, x, width, color) {
+    this.x = x;
+    this.width = width;
+    this.color = color;
+    this.vm = vm;
+    this.game = game;
+    this.keyObj = keyObj;
+    this.key = key;
+    this.ctx = vm.ctx;
+    this.canvas = vm.canvas;
+    this.percentage = 0;
+    this.isHoldNote = keyObj.h && keyObj.h[key];
+    this.isUserHolding = false;
+    this.isHoldingDone = false;
 
-  this.y = 0;
+    this.y = 0;
+    this.singleNoteHeight = 10;
 
-  // modulate speed, ref https://www.viget.com/articles/time-based-animation/
-  this.now = Date.now();
-  this.delta = 0;
-  this.then = 0;
-  this.timeStarted = this.now;
+    // modulate speed, ref https://www.viget.com/articles/time-based-animation/
+    this.now = Date.now();
+    this.delta = 0;
+    this.then = 0;
+    this.timeStarted = this.now;
+  }
 
-  this.setDelta = function () {
+  setDelta() {
     if (this.then === 0) this.then = this.now;
```

Comment:
Would it be possible for time dilation? Really really edge case but who knows...

Reply:
Actually this delta is a must I think... Because I don't know the exact paint rate of the canvas animation frame function, so this might add more or less pixels than required, which is a big issue because I need the note to arrive at the line in exact milliseconds, don't know if there is any better way to do this.